### DISCUSSION TITLE
ESP-NOW - TasMesh Code

### DISCUSSION THREAD

**Discussion by ojundi** [arendst/Tasmota#11939] [Ideas]
Hello,

I am wondering why was ESP-now code named "Tasmesh" was removed from the repo and was basically wiped clean with no mention. 

https://github.com/arendst/Tasmota/commit/e3d3e76d6169baeb0b804114f843de94e77bc738
Thanks,
Oz

**Comment by sfromis** [comment]
Looks like a short-lived playground thing, since the experimental "proof of concept" code was removed the day after it was added. My guess is that it did not work, and the developer decided against pursuing the idea in the repository.

**Reply by ascillato** [reply]
Seems that it wasn't intended to be added to Tasmota. The commit that have added that to the codebase was between other commits in the PR https://github.com/arendst/Tasmota/pull/9588. That PR wasn't about Mesh at all. So, seems that the Mesh code is a WIP of @Staars and not ready.

**Reply by Staars** [reply]
That was a mistake from my side doing a commit from an unclean repo.

The mesh code did run pretty well, but I could not really attract other developers in the discord channel (with one exeption), so I put in on hold. The problem with such a potential core function is the design and integration into the main project, not the coding itself.

The basic idea is to turn most of the Tasmota devices in your network into a node and then communicate to central ESP32 (as a broker or gateway), which is the only necessary WiFi device. All nodes form a mesh automatically, so you rarely will see range problems. Battery powered devices need only an uptime of about a second to send a message and then go to sleep again.
This is all realized by routing the usual MQTT messages via ESP-now, thus no driver updates are needed.

**Reply by chaveiro** [reply]
So will you add it again to the main repo?

**Reply by ojundi** [reply]
There are several examples of new products out there using the mesh technology now.  I assume since the ESP-now driver is being used ESP8266 will be able to also act as a gateway as well with more limitations. 

I appreciate if this moves forward or to a location to where the community can contribute.

**Reply by Staars** [reply]
We are still working on it and currently are testing it in real-world scenarios. At least it is still working with the current development branch. It is not forgotten.

**Reply by ojundi** [reply]
if i can be of assistance on this.. i have about 850 sonoff mini / ESP8266 left in stock.. i can do some aggressive testing if that helps..

**Comment by seedsofhope** [comment]
What's the official word on the mesh? Will it be included in tasmota or not?

**Reply by ascillato2** [reply]
Now, It has been included in Tasmota https://github.com/arendst/Tasmota/commit/16f6f26aba72945c2ef220e685d8162fff63891e

**Comment by arendst** [comment]
I received a note from the original owners of the WIP that they are unable to work on future issues on this WIP.

I also took the liberty of scanning the code and made it work on the latest Tasmota release. As the original code was a WIP it will now be a POC as I needed to make some important changes and additions to make it at least user safe without too many exceptions.

I'll release it in the next days but as said it still needs a lot of polishing as important features are lacking:

- no node receive queue and broker send queue so no mqtt commands with data over 100 characters. Full mqtt response sizes are supported though
- no command persistance so use rules for the short term

In addition it's slow compared to mqtt over wifi but for simple power commands it's very useful. If I drop the packet time to 10ms it allows for OTA over MQTT of an 300k binary within 11 minutes :-) without dropped packages so in core it's useful.

Stay tuned.

**Reply by effelle** [reply]
I think OTA is out of the scope of a mesh due to the data rate and the limited payload, we used 160 char + headers to be sure to be under 200 chars total and that's a limit for the esp8266, but I got your point. I'm also admit I didn't think on OTA over Mesh at all. :-)

**Reply by digiblur** [reply]
Awesome stuff!  Nice job everyone!

**Reply by ascillato** [reply]
Now, It has been included in Tasmota https://github.com/arendst/Tasmota/commit/16f6f26aba72945c2ef220e685d8162fff63891e

**Reply by Staars** [reply]
Wow, did not expect this to happen.
I can agree, that I never even thought about OTA via ESP-now too and yes, this driver is very unfinished.

**Reply by ojundi** [reply]
seriously wow!!.. thats for sure why this repo is a winner!!.. Go Tasmota.. 

Will be doing some testing soon!!

**Reply by ojundi** [reply]
I also hope that this will be the basis for something like Amazon sidewalk in the near future!!

**Comment by nagyrobi** [comment]
Some questions in my mind...

- Is the number of encrypted peers set to 10 here [too](https://randomnerdtutorials.com/esp-now-esp32-arduino-ide/)? (how many encrypted esp8266 peers could connect to an esp32 gateway?)
- can the esp32-based devices equipped with a (poe) ethernet port be supported? Like using the wireless interface for ESP-NOW only, and the Ethernet port to talk to MQTT. Perhaps could have a positive impact on the previous question too.
- would it be possible to support multiple esp32 gateways to avoid Single Point of Failure. Currently each Tasmota connects directly to the network and if one breaks, only the related functionality is lost. With TasMesh if the esp32 gateway happens to break, we loose everything behind. If the nodes could contact a secondary/tertiary gateway after some timeout we could have some nice redundancy.

**Reply by Staars** [reply]
Encryption is handled entirely in the driver thus the limit of the ESP-SDK does not apply here.
But every node (and the broker too) can only broadcast the messages to max. 20 receiver nodes, which is a limitation of the SDK. 
In theory the driver can handle a grid of many more nodes, but you must mix different sets of node addresses per node.
Let's say you have an ESP32 as a  broker with 40 nodes (simplified addresses 1-40). The broker can only broadcast to node 1-20. You would have to make sure, that i.e. node 1 knows the addresses of node 2-20, node 2 knows node 11-30 and node 3 knows 21-40, ....

I know, that this could have been solved in many different ways, but I chose this path back in the days.

BTW, real world testing of the mesh functionality was really hard, because if you do not live in big house, chances are quite good, that the ESP32 can reach every ESP8266-node directly. Thus it would not really form a mesh.

**Reply by Staars** [reply]
Yes, Single point of failure is an ugly problem and probably one of the main reasons, why we hesitated to do a PR earlier.

**Reply by nagyrobi** [reply]
In principle with two gateways you'd actually have the second gateway act much like a node. It should watch for the main gateway periodically and if that is dead connect to mqtt and do the work instead etc...

**Reply by effelle** [reply]
I did some tests with an ESP32 and a ETH module (LAN8720). 
Worked as expected for normal operations but I didn't had the time to change the code to manage the wifi connection to act as an access point for the mesh network only. 
The idea was to enable ETH to add the ability to choose the wifi channel for the mesh and eventually avoid some interference.

**Reply by Cartman18** [reply]
@effelle Great Work! 
Please stay tuned on "SmartHome with Tasmota but without WLAN".
We switch off WiFi over night (WAF factor :) ) so the usual tasmota devices are not reachable any more. Your 
"LAN<->ESP32<->TasMesh" solution would be great in this case. 
Next to all the other advantages of TasMesh (power consumption, range, .. )
Thx in advance!

**Reply by FaserF** [reply]
> @effelle Great Work! Please stay tuned on "SmartHome with Tasmota but without WLAN". We switch off WiFi over night (WAF factor :) ) so the usual tasmota devices are not reachable any more. Your "LAN<->ESP32<->TasMesh" solution would be great in this case. Next to all the other advantages of TasMesh (power consumption, range, .. ) Thx in advance!

But how stays your "Broker" online if you switch off Wifi at night? I mean if you turn off wifi, this tasmesh wont help or am I wrong?
I would love to turn off wifi too at night.

**Comment by ortegafernando** [comment]
Great!!! It will be perfect also to use MQTT with SSL. As ESP8266 can't, we can use esp-now to send trafic to our ESP32 "broker/bridge", and then, thought SECURE MQTT, send it to our broker.
Thanks a lot.

**Comment by FaserF** [comment]
Sorry if ese are dump questions, but would the following be possible in the future?
1. I am running Homeassistant on my Raspberry Pi via LAN. Can I somehow use the raspberry Pi with its wifi as the TasMesh Broker?
Why I am asking it: That way I wouldnt need wifi for my tasmota devices at all, as in the end the tasmota devices would just use the LAN of the raspberry pi.


2. And the only docs about TasMesh are here? https://gist.github.com/Staars/42f8b41ddc593d25db868b30b8c28721
I wasnt able to find any other docs and how to's to set it up.

3. Do I need to use one ESP32 ONLY for TasMesh or can this device itself control something and be the TasMesh Broker at once?

Thanks in advance.

**Reply by Staars** [reply]
It is unlikely, that a non-ESP-device will be able to really support ESP-NOW in the future, although it might technically possible (I assume, someone would have to modify the network driver of the raspberry at  a lower level).
The „broker“ can do other things, as long as enough resources are left on the device. There is no limit hardcoded in the driver.

**Reply by Jason2866** [reply]
> Sorry if ese are dump questions, but would the following be possible in the future?
> 
> 1. I am running Homeassistant on my Raspberry Pi via LAN. Can I somehow use the raspberry Pi with its wifi as the TasMesh Broker?
>    Why I am asking it: That way I wouldnt need wifi for my tasmota devices at all, as in the end the tasmota devices would just use the LAN of the raspberry pi.
> 2. And the only docs about TasMesh are here? https://gist.github.com/Staars/42f8b41ddc593d25db868b30b8c28721
>    I wasnt able to find any other docs and how to's to set it up.
> 3. Do I need to use one ESP32 ONLY for TasMesh or can this device itself control something and be the TasMesh Broker at once?
> 
> Thanks in advance.

Here is the docu https://github.com/arendst/Tasmota/blob/development/info/xdrv_57_tasmesh.md

**Comment by barbudor** [comment]
You may switch off your AP at night but that doesn't switch off WiFi. How do you think ESP-NOW mesh is working? It's using wifi mesh.
So if your goal is to reduce exposition to wifi waves during sleep, you don't.

**Reply by ascillato** [reply]
Exactly. A mesh network of Tasmota devices will use much more energy and will irradiate much more power to the air than the regular Wi-Fi router + devices in Station mode. When Tasmota is in Mesh, in reality it is STA + AP mode, so the sleep feature is disabled (because of being in AP mode) and the power consumption and power irradiation are much more higher.

So, if you are trying to use mesh for lowering energy or for irradiating less power to the air, you are doing just the opposite.

Mesh is for expanding the range of a Wi-Fi Network.

If you want your devices to STOP using Wi-Fi at night, it is possible using a simple timer and a rule. But be aware that while the Wi-Fi of the device is OFF, you won't be able to control it from the network. Only from the buttons or switches of the device itself.

**Reply by FaserF** [reply]
> Exactly. A mesh network of Tasmota devices will use much more energy and will irradiate much more power to the air than the regular Wi-Fi router + devices in Station mode. When Tasmota is in Mesh, in reality it is STA + AP mode, so the sleep feature is disabled (because of being in AP mode) and the power consumption and power irradiation are much more higher.
> 
> So, if you are trying to use mesh for lowering energy or for irradiating less power to the air, you are doing just the opposite.
> 
> Mesh is for expanding the range of a Wi-Fi Network.
> 
> If you want your devices to STOP using Wi-Fi at night, it is possible using a simple timer and a rule. But be aware that while the Wi-Fi of the device is OFF, you won't be able to control it from the network. Only from the buttons or switches of the device itself.

I thought it saves battery?
„ Power consumption of the nodes will be reduced significantly allowing better battery powered projects with TASMOTA“

https://github.com/arendst/Tasmota/blob/development/info/xdrv_57_tasmesh.md

**Reply by ascillato** [reply]
No, it uses much more energy. For saving battery, the approach is totally different like being in STA mode only and with deepsleep intervals enabled.

**Comment by jkdobro** [comment]
Maybe this EnigmaIOT:
https://github.com/gmag11/EnigmaIOT

**Comment by ferbulous** [comment]
Hi, sorry to bump this old discussion thead.
I was under the assumption that tasmesh would continue working if the wifi/mqtt broker is down.
I did a simple test with 2 esp8266 devices (light switch & bulb) as nodes & esp32 devboard as meshbroker with a simple rule that just toggles the light on or off with a button press and the main server broker turned off

```
rule1 
on power1#state do publish cmnd/light/power %state% endon
```
But it wasn't toggling the light bulb until I had the main server broker online again.

```
06:52:08.922 MQT: Attempting connection...
06:52:08.930 MQT: Connect failed to 192.168.1.40:1883, rc -2. Retry in 10 sec
06:52:19.924 MQT: Attempting connection...
06:52:19.935 MQT: Connect failed to 192.168.1.40:1883, rc -2. Retry in 20 sec
06:52:23.722 RSL: RESULT = {"Command":"Unknown"}
06:52:25.823 RSL: RESULT = {"POWER":"ON"}
06:52:26.023 RSL: POWER = ON
06:52:26.173 RSL: power = 1
06:52:29.723 RSL: RESULT = {"POWER":"OFF"}
06:52:29.923 RSL: POWER = OFF
06:52:30.123 RSL: power = 0
06:52:32.023 RSL: RESULT = {"POWER":"ON"}
06:52:32.242 RSL: POWER = ON
06:52:32.373 RSL: power = 1
06:52:36.423 RSL: RESULT = {"POWER":"OFF"}
06:52:36.673 RSL: POWER = OFF
06:52:36.923 RSL: power = 0
06:52:40.873 RSL: RESULT = {"POWER":"ON"}
06:52:40.925 MQT: Attempting connection...
06:52:40.935 MQT: Connect failed to 192.168.1.40:1883, rc -2. Retry in 30 sec
06:52:41.073 RSL: POWER = ON
06:52:41.223 RSL: power = 1
06:52:44.423 RSL: RESULT = {"POWER":"OFF"}
06:52:44.673 RSL: POWER = OFF
06:52:44.923 RSL: power = 0
06:53:11.926 MQT: Attempting connection...
06:53:11.944 MQT: Connected
06:53:11.948 MQT: tele/esp32meshtest/LWT = Online (retained)
06:53:11.951 MQT: cmnd/esp32meshtest/POWER = 
06:53:11.954 MSH: Reconnect topic switch_test
06:53:11.956 MSH: Reconnect topic light_test
06:53:13.183 MQT: stat/esp32meshtest/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/release/tasmota32.bin","RestartReason":"RTC Watch dog reset digital core and rtc module","Uptime":"0T00:26:25","StartupUTC":"2024-08-05T05:26:48","Sleep":50,"CfgHolder":4617,"BootCount":20,"BCResetTime":"2024-05-29T15:18:32","SaveCount":413}}
06:53:13.236 MQT: stat/esp32meshtest/STATUS11 = {"StatusSTS":{"Time":"2024-08-05T06:53:13","Uptime":"0T00:26:25","UptimeSec":1585,"Heap":131,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":21,"MqttCount":13,"Berry":{"HeapUsed":4,"Objects":42},"POWER":"OFF","Wifi":{"AP":1,"SSId":"my_wifi","BSSId":"9C:A2:F4:37:D5:E1","Channel":9,"Mode":"HT20","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}}
06:53:13.284 MQT: stat/esp32meshtest/RESULT = {"Time":"2024-08-05T06:53:13","Uptime":"0T00:26:25","UptimeSec":1585,"Heap":133,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":22,"MqttCount":13,"Berry":{"HeapUsed":4,"Objects":42},"POWER":"OFF","Wifi":{"AP":1,"SSId":"my_wifi","BSSId":"9C:A2:F4:37:D5:E1","Channel":9,"Mode":"HT20","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}
06:53:13.336 MQT: stat/esp32meshtest/STATUS10 = {"StatusSNS":{"Time":"2024-08-05T06:53:13","MESH":{"channel":9,"nodes":2,"MAC":["40:91:51:4C:D6:E9","C8:C9:A3:6A:97:61"]}}}
06:53:22.377 MQT: stat/switch_test/RESULT = {"POWER":"ON"}
06:53:22.677 MQT: stat/switch_test/POWER = ON
06:53:22.926 MQT: cmnd/light_test/power = 1
06:53:23.276 MQT: stat/light_test/RESULT = {"POWER":"ON"}
06:53:23.394 MQT: stat/light_test/POWER = ON
06:53:27.077 MQT: stat/switch_test/RESULT = {"POWER":"OFF"}
06:53:27.278 MQT: stat/switch_test/POWER = OFF
06:53:27.427 MQT: cmnd/light_test/power = 0
06:53:27.577 MQT: stat/light_test/RESULT = {"POWER":"OFF"}
06:53:27.728 MQT: stat/light_test/POWER = OFF

06:53:28.927 MQT: Attempting connection...
06:53:28.951 MQT: Connected
06:53:28.955 MQT: tele/esp32meshtest/LWT = Online (retained)
06:53:28.958 MQT: cmnd/esp32meshtest/POWER = 
06:53:28.962 MSH: Reconnect topic switch_test
06:53:28.964 MSH: Reconnect topic light_test
06:53:30.084 MQT: stat/esp32meshtest/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/release/tasmota32.bin","RestartReason":"RTC Watch dog reset digital core and rtc module","Uptime":"0T00:26:42","StartupUTC":"2024-08-05T05:26:48","Sleep":50,"CfgHolder":4617,"BootCount":20,"BCResetTime":"2024-05-29T15:18:32","SaveCount":413}}
06:53:30.135 MQT: stat/esp32meshtest/STATUS11 = {"StatusSTS":{"Time":"2024-08-05T06:53:30","Uptime":"0T00:26:42","UptimeSec":1602,"Heap":131,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":21,"MqttCount":14,"Berry":{"HeapUsed":4,"Objects":42},"POWER":"OFF","Wifi":{"AP":1,"SSId":"my_wifi","BSSId":"9C:A2:F4:37:D5:E1","Channel":9,"Mode":"HT20","RSSI":52,"Signal":-74,"LinkCount":1,"Downtime":"0T00:00:03"}}}
06:53:30.186 MQT: stat/esp32meshtest/RESULT = {"Time":"2024-08-05T06:53:30","Uptime":"0T00:26:42","UptimeSec":1602,"Heap":133,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":22,"MqttCount":14,"Berry":{"HeapUsed":4,"Objects":42},"POWER":"OFF","Wifi":{"AP":1,"SSId":"my_wifi","BSSId":"9C:A2:F4:37:D5:E1","Channel":9,"Mode":"HT20","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}
06:53:30.238 MQT: stat/esp32meshtest/STATUS10 = {"StatusSNS":{"Time":"2024-08-05T06:53:30","MESH":{"channel":9,"nodes":2,"MAC":["40:91:51:4C:D6:E9","C8:C9:A3:6A:97:61"]}}}
06:53:30.878 MQT: stat/switch_test/RESULT = {"POWER":"ON"}
06:53:31.077 MQT: stat/switch_test/POWER = ON
06:53:31.228 MQT: cmnd/light_test/power = 1
06:53:31.628 MQT: stat/light_test/RESULT = {"POWER":"ON"}
06:53:31.777 MQT: stat/light_test/POWER = ON

06:53:32.927 MQT: Attempting connection...
06:53:32.945 MQT: Connected
06:53:32.949 MQT: tele/esp32meshtest/LWT = Online (retained)
06:53:32.952 MQT: cmnd/esp32meshtest/POWER = 
06:53:32.955 MSH: Reconnect topic switch_test
06:53:32.957 MSH: Reconnect topic light_test
06:53:34.041 MQT: stat/esp32meshtest/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/release/tasmota32.bin","RestartReason":"RTC Watch dog reset digital core and rtc module","Uptime":"0T00:26:46","StartupUTC":"2024-08-05T05:26:48","Sleep":50,"CfgHolder":4617,"BootCount":20,"BCResetTime":"2024-05-29T15:18:32","SaveCount":413}}
06:53:34.093 MQT: stat/esp32meshtest/STATUS11 = {"StatusSTS":{"Time":"2024-08-05T06:53:34","Uptime":"0T00:26:46","UptimeSec":1606,"Heap":132,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":25,"MqttCount":15,"Berry":{"HeapUsed":4,"Objects":42},"POWER":"OFF","Wifi":{"AP":1,"SSId":"my_wifi","BSSId":"9C:A2:F4:37:D5:E1","Channel":9,"Mode":"HT20","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}}
06:53:34.141 MQT: stat/esp32meshtest/RESULT = {"Time":"2024-08-05T06:53:34","Uptime":"0T00:26:46","UptimeSec":1606,"Heap":132,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":26,"MqttCount":15,"Berry":{"HeapUsed":4,"Objects":42},"POWER":"OFF","Wifi":{"AP":1,"SSId":"my_wifi","BSSId":"9C:A2:F4:37:D5:E1","Channel":9,"Mode":"HT20","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}
06:53:34.192 MQT: stat/esp32meshtest/STATUS10 = {"StatusSNS":{"Time":"2024-08-05T06:53:34","MESH":{"channel":9,"nodes":2,"MAC":["40:91:51:4C:D6:E9","C8:C9:A3:6A:97:61"]}}}
06:53:34.484 MQT: stat/switch_test/RESULT = {"POWER":"OFF"}
06:53:34.784 MQT: stat/switch_test/POWER = OFF
06:53:35.034 MQT: cmnd/light_test/power = 0
06:53:35.333 MQT: stat/light_test/RESULT = {"POWER":"OFF"}
06:53:35.433 MQT: stat/light_test/POWER = OFF
```

Am I correct to assume that if main server broker is down then it's unable to publish anything?

**Reply by sfromis** [reply]
You can't use MQTT without a broker, Tasmesh changes nothing about that. And while Tasmesh internally uses ESP-NOW, it is not intended to make your setup independent of having IP networking from the gateway. (I avoided the confusion of using the word "broker" for both, as the Tasmesh gateway is not a MQTT broker).

**Reply by ferbulous** [reply]
I see, I assumed that you could still publish just with the esp32 tasmesh broker

**Comment by Styne13** [comment]
Hi,

I have another question regarding tasmesh and the „Node last message“ or mesh interval.

My setup includes ESP32 as meshbroker and ESP8266 with DS18B20 temperatur sensor as node. After I configured MeshChannel and MeshIntetval on both nodes, start the MeshBroker and connect to it with MeshNode,
I only get the connection once.
The „Node last message“ is updated on the UI of the ESP32, but I won‘t get any further messages/updates.

Do I have to configure anything else? My thought was that I do not have to configure MQTT on the mesh nodes but only on the mesh broker.

Thanks and BR

Marcel

**Reply by sfromis** [reply]
Not sure how much you mean by "My thought was that I do not have to configure MQTT on the mesh nodes" when the whole concept is about transporting MQTT messages via the mesh network. As the [TasMesh docs](https://github.com/arendst/Tasmota/blob/development/info/xdrv_57_tasmesh.md) says:
> The nodes will send their MQTT-messages back to the broker via ESP-NOW.

Simply changing an ESP8266 device already configured normally for MQTT to be a TasMesh node worked right away, with me seeing the node MQTT payloads on both the ESP32 mesh gateway and my MQTT broker.

**Reply by Styne13** [reply]
I‘m sorry. It was my fault to missinterpret the documentation that I do not have to configure the mqtt broker connection.

Thanks for clarification! Afer doing so and switching to MeshNode afterwards, everything works fine. :)

**Reply by Styne13** [reply]
I heared that ESP-NOW would have higher range than using normal station mode.

What are your experiences using tasmesh?
For me using standard pcb antennas it seems to be a lower range, because I can‘t connet the node through two walls.

**Reply by Staars** [reply]
The main idea of Tasmesh is the mesh. Thus you can extend the range with more nodes in between.
But besides in my personal tests back in the days I had problems to „cut the the connection“ between one broker ESP32 and an ESP8266 node to check, if a „relay node“ is actually working.
As always YMMV.

**Reply by sfromis** [reply]
When it comes to range, remember that most ESP devices may not have as good Wifi hardware (including antenna and signal strength) as dedicated router/AP devices. If you want a more realistic range comparison, you could try having one case being TasMesh, and the other being the Tasmota range extender feature. But yes, with a much lower data rate, ESP-Now might survive a bit lower signal strength, on the exact same hardware.

**Reply by Styne13** [reply]
Ok, thank you guys for your ratings.
I will also check out the range extender feature and compare for my usecase.
